File: day07/src/main.rs
use std::cell::RefCell;
use std::cmp::min;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::rc::Rc;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "/Users/ecem/CLionProjects/adventofcode2022/day07/day7.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

struct Dir
{
    name: String,
    files: Vec<(String, i64)>,
    dirs: Vec<Rc<RefCell<Dir>>>,
}

impl Dir {
    fn new(name: &str) -> Rc<RefCell<Dir>> {
        Rc::new(RefCell::new(Dir { name: name.to_string(), files: vec![], dirs: vec![] }))
    }
}

fn parse() -> Rc<RefCell<Dir>> {
    let mut stack = vec![Dir::new("/")];

    for line in read() {
        let tokens = line.split(' ').collect::<Vec<&str>>();

        match tokens[0] {
            "$" => {
                match tokens[1] {
                    "cd" => {
                        match tokens[2] {
                            "/" => { stack.drain(1..stack.len()); }
                            ".." => { stack.pop(); }
                            dirname => {
                                let last_ptr = stack.last().unwrap().clone();
                                let last = last_ptr.borrow();
                                let found = last.dirs.iter().find(|x| {
                                    x.borrow().name == dirname
                                }).unwrap();
                                stack.push(found.clone());
                            }
                        };
                    }
                    "ls" => {
                        let last_ptr = stack.last().unwrap().clone();
                        let mut last = last_ptr.borrow_mut();
                        last.files.clear();
                        last.dirs.clear();
                    }
                    c => panic!("unknown command {}", c)
                }
            }
            "dir" => {
                let last_ptr = stack.last().unwrap().clone();
                let mut last = last_ptr.borrow_mut();
                last.dirs.push(Dir::new(tokens[1]));
            }
            filesize => {
                let last_ptr = stack.last().unwrap().clone();
                let mut last = last_ptr.borrow_mut();
                last.files.push((tokens[1].to_string(), filesize.parse::<i64>().unwrap()));
            }
        }
    }

    return stack.first().unwrap().clone();
}

fn traverse<F: FnMut(&Rc<RefCell<Dir>>)>(node: &Rc<RefCell<Dir>>, func: &mut F) {
    func(node);

    for dir in &node.borrow().dirs {
        traverse(dir, func);
    }
}

fn getsize(node: &Rc<RefCell<Dir>>) -> i64 {
    let mut result: i64 = 0;
    for (_, size) in &node.borrow().files {
        result += size;
    }
    for dir in &node.borrow().dirs {
        result += getsize(dir);
    }

    result
}

fn part1() {
    let root = parse();

    let mut total: i64 = 0;
    traverse(&root, &mut |dir| {
        let size = getsize(dir);
        if size <= 100000 {
            total += size;
        }
    });

    println!("Part1: {}", total);
}

fn part2() {
    let root = parse();

    let used_size = getsize(&root);
    let disk_space = 70_000_000;
    let update_size = 30_000_000;
    let space_needed = update_size - (disk_space - used_size);

    let mut result: i64 = disk_space;
    traverse(&root, &mut |dir| {
        let size = getsize(dir);
        if size >= space_needed {
            result = min(result, size);
        }
    });

    println!("Part2: {}", result);
}

fn main() {
    part1();
    part2();
}

File: day07/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>
#include <unordered_set>
#include <unordered_map>
#include <map>

std::vector<std::string> split(const std::string& str)
{
    std::istringstream split(str);
    std::vector<std::string> tokens;
    for (std::string each; std::getline(split, each, ' '); tokens.push_back(each));
    return tokens;
}

auto read() {
    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}


struct Dir
{
    Dir(std::string n): name(n){}
    std::string name;
    std::vector<std::pair<std::string, int>> files;
    std::vector<std::shared_ptr<Dir>> dirs;
};

int64_t get_size(const std::shared_ptr<Dir>& dir)
{
    int64_t result = 0;
    for(const auto& child: dir->dirs)
    {
        result+=get_size(child);
    }
    for(const auto& file: dir->files)
    {
        result+=file.second;
    }
    return result;
}

void traverse(const std::shared_ptr<Dir>& dir, std::function<void(const std::shared_ptr<Dir>&)> function)
{
    function(dir);
    for(const auto& child: dir->dirs)
        traverse(child, function);
}

std::shared_ptr<Dir> parse()
{
    std::vector<std::shared_ptr<Dir>> stack = {std::make_shared<Dir>("/")};

    for(const auto& line: read())
    {
        const auto tokens = split(line);
        if(tokens[0]=="$") // command
        {
            if(tokens[1]=="cd")
            {
                if(tokens[2]=="/")
                {
                    stack = {stack.front()};
                }
                else if(tokens[2]=="..")
                {
                    stack.pop_back();
                }
                else
                {
                    const auto& children = stack.back()->dirs;
                    auto it = std::find_if(children.begin(), children.end(),[&](const std::shared_ptr<Dir>& dir){
                        return dir->name==tokens[2];
                    });
                    assert(it!=children.end());
                    stack.push_back(*it);
                }
            }
            else if(tokens[1]=="ls")
            {
                stack.back()->dirs.clear();
                stack.back()->files.clear();
            }
        }
        else if(tokens[0]=="dir")
        {
            auto& children = stack.back()->dirs;
            auto it = std::find_if(children.begin(), children.end(),[&](const std::shared_ptr<Dir>& dir){
                return dir->name==tokens[1];
            });
            assert(it==children.end());
            children.push_back(std::make_shared<Dir>(tokens[1]));
        }
        else
        {
            stack.back()->files.push_back({tokens[1], std::stoi(tokens[0])});
        }
    }
    return stack.front();
}

void part1() {
    auto root = parse();
    int64_t total = 0;
    traverse(root, [&total](const std::shared_ptr<Dir>& dir){
        int64_t current_size = get_size(dir);
        if(current_size<=100000)
            total+=current_size;
    });
    std::cout << "Part1: " << total << std::endl;
}

void traverse2(const std::shared_ptr<Dir>& dir, int64_t& result, int64_t search_for)
{
    int64_t current_size = get_size(dir);
    if(current_size>=search_for)
        result=std::min(result, current_size);

    for(const auto& child: dir->dirs)
    {
        traverse2(child, result, search_for);
    }

}

void part2() {
    auto root = parse();

    int64_t result = std::numeric_limits<int32_t>::max();
    const int64_t disk_space = 70000000;
    const int64_t update_size = 30000000;
    auto space_needed = update_size-disk_space+get_size(root);

    traverse(root, [&](const std::shared_ptr<Dir>& dir){
        int64_t current_size = get_size(dir);
        if(current_size>=space_needed)
            result=std::min(result, current_size);
    });
    std::cout << "Part2: " << result << std::endl;
}

int main() {
    part1();
    part2();
    return 0;
}
